libcontainer/nsenter/cloned_binary.c

```diff
@@ -192,49 +197,155 @@ static int fetchve(char ***argv)
 	return -EINVAL;
 }
 
+#define EFD_NONE     0x00
+#define EFD_MEMFD    0x01
+#define EFD_TMPFILE  0x02
+#define EFD_FILE     0x04
+#define EFD_ALL      (EFD_MEMFD | EFD_TMPFILE | EFD_FILE)
+
+static int make_execfd(int *fdtype, int allowed)
+{
+	int fd = -1;
+	char template[PATH_MAX] = {0};
+	char *prefix = secure_getenv("_LIBCONTAINER_STATEDIR");
+
+	if (!prefix || *prefix != '/')
+		prefix = "/tmp";
+	if (snprintf(template, sizeof(template), "%s/runc.XXXXXX", prefix) < 0)
+		return -1;
+
+	/*
+	 * Try memfd first, it's much nicer since it's easily detected thanks to
+	 * sealing and also doesn't require assumptions like /tmp.
+	 */
+	if (allowed & EFD_MEMFD) {
+		*fdtype = EFD_MEMFD;
+		fd = memfd_create(RUNC_MEMFD_COMMENT, MFD_CLOEXEC | MFD_ALLOW_SEALING);
+		if (fd >= 0)
+			return fd;
+		if (errno != ENOSYS)
```

Comment:
What if os has `memfd_create` syscall and without seal support (I know that can happen on centos 7.2 with `3.10.0-327.36.3` kernel). It will return `EINVAL` without retry other ways.

By the way, is it reasonable to just allow use `memfd_create` without such sealing flags?

Reply:
I can make this check `errno != EINVAL` too.

> By the way, is it reasonable to just allow use memfd_create without such sealing flags?

No, because while we don't need them to protect against overwriting, we need working sealing support so that we can detect whether we are a cloned binary.